articles/115_topic_and_service_name_mapping.md

```diff
@@ -0,0 +1,317 @@
+A relative name does not begin with a forward slash (`/`) and does respect the namespace of the node which created them.
+
+Relative names are appended to the namespace of the node which creates them.
+For example, if the node is put into a namespace `/ping/pong` and the topic or service name is `foo/bar`, then the absolute name will become `/ping/pong/foo/bar`.
+However, if the topic or service name is `/foo/bar`, then the absolute name will stay just `/foo/bar`, ignoring the node's namespace.
+
+### Name Tokens
+
+Name token's are the strings between the namespace delimiters, e.g. the tokens of the topic or service `/foo/bar/baz` are `foo`, `bar`, and `baz`.
+
+Topic and service name tokens:
+
+- must not be empty, e.g. the name `//bar` is not allowed
+
+  - rational: it removes the chance for accidental `//` from concatenation; also if the implementation chooses to use a trailing underscore (`_`) then it prevents `foo_/_bar` and `foo//bar` from being ambiguous
+
+- may use alphanumeric characters (`[0-9|a-z|A-Z]`), or an underscore (`_`)
+
+- must not start with numeric characters (`[0-9]`)
+
+- must not end with a single underscore (`_`)
+
+  - rational: if tokens are allowed to end with and start with `_` then `foo_/bar` is indistinguishable from `foo/_bar` if the `/` is replaced with `__`, i.e. both result in `foo___bar`
+
+- must not be a single underscore (`_`)
+
+  - rational: this is a special case of "must not end with an underscore"
+
+- must not have two or more underscores (`__`) repeated anywhere
+
+  - rational: this provides the implementation with a safe to use delimiter
+
+- may be a single tilde character (`~`)
+
+### Private Namespace Substitution Character
+
+The special single character token `~` will be replaced with a namespace snippet that is a concatenation of the namespace for the node and the node name.
+For example, for a node `node1` in a namespace `/foo` would result in `~` being replaced with `/foo/node1`.
+For another example, a node `node1` in a namespace `foo/bar` would result in `~` being replaced with `foo/bar/node1`.
+It may be used in any token position.
+Any resulting empty namespaces which result from expanding the `~` will be collapsed.
+For example, when `~` expands to `/foo/node1` and the name `ping/~/pong` is given, it will expand to `ping/foo/node1/pong` and not to `ping//foo/node1/pong`.
+Here is a table with some example expansions:
+
+| **Input Name** | Node: `my_node` NS: none | Node: `my_node` NS: `/foo`     |
+|----------------|--------------------------|--------------------------------|
+| `ping`         | *`/ping`*                | *`/foo/ping`*                  |
+| `/ping`        | *`/ping`*                | *`/ping`*                      |
+| `~`            | *`/my_node`*             | *`/foo/foo/my_node`*           |
+| `~/ping`       | *`/my_node/ping`*        | *`/foo/foo/my_node/ping`*      |
```

Comment:
I like the idea of having a subsitution token such as `${ns}` instead of "~". However I worry about the following case:
```bash
# Somewhere this is defined:
export ns="some_value"

# The user now tries to manually instantiate a node passing the topic as parameter:
my_node my_parameter_with_namespace:=${ns}/some_topic
# bash would interpret it as:
my_node my_parameter_with_namespace:=some_value/some_topic
# but this could be avoided with:
my_node 'my_parameter_with_namespace:=${ns}/some_topic'
# which is a little cumbersome and not everyone knows about it.
```
I would go with what @wjwwood suggested with "{ns}" or "%{}" that would not be a source of very obscure and hard to debug errors.